compiler/hash-typecheck/src/fmt.rs

```diff
@@ -0,0 +1,440 @@
+                        write!(f, "> -> ")?;
+                        self.fmt_ty(f, *general_return_ty, &Cell::new(false))?;
+
+                        if let Some(case) = cases.first() {
+                            write!(f, " => ")?;
+                            let is_atomic = Cell::new(false);
+                            self.fmt_value(f, case.return_value, &is_atomic)?;
+                        }
+
+                        // We assume only the first case is the base one
+                        // @@TODO: refine this to show all cases
+                        Ok(())
+                    }
+                }
+            }
+            Value::AppTyFn(app_ty_fn) => {
+                is_atomic.set(true);
+                self.fmt_app_ty_fn(f, app_ty_fn)
+            }
+            Value::Var(var) => {
+                is_atomic.set(true);
+                write!(f, "{}", var.name)
+            }
+            Value::Merge(values) => {
+                is_atomic.set(false);
+                for (i, value) in values.iter().enumerate() {
+                    let value_is_atomic = Cell::new(false);
+                    let value_fmt = format!(
+                        "{}",
+                        value
+                            .for_formatting_with_atomic_flag(self.global_storage, &value_is_atomic)
+                    );
+
+                    if !value_is_atomic.get() {
+                        write!(f, "(")?;
+                    }
+                    write!(f, "{}", value_fmt)?;
+                    if !value_is_atomic.get() {
+                        write!(f, ")")?;
+                    }
+
+                    if i != values.len() - 1 {
+                        write!(f, " ~ ")?;
+                    }
+                }
+                Ok(())
+            }
+            Value::Unset => {
+                is_atomic.set(true);
+                write!(f, "{{unset}}")
+            }
+            Value::ModDef(_) => todo!(),
+            Value::NominalDef(_) => todo!(),
+        }
+    }
+
+    pub fn fmt_unresolved(
+        &self,
+        f: &mut fmt::Formatter,
+        UnresolvedTy { resolution_id }: &UnresolvedTy,
+    ) -> fmt::Result {
+        write!(f, "{{unresolved({:?})}}", resolution_id,)
+    }
+
+    pub fn fmt_nominal_def(
+        &self,
+        f: &mut fmt::Formatter,
+        nominal_def_id: NominalDefId,
+        is_atomic: &Cell<bool>,
+    ) -> fmt::Result {
+        is_atomic.set(true);
+        match self.global_storage.nominal_def_store.get(nominal_def_id) {
+            NominalDef::Struct(StructDef {
+                name: Some(name), ..
+            })
+            | NominalDef::Enum(EnumDef {
+                name: Some(name), ..
+            }) => {
+                write!(f, "{}", name)
+            }
+            NominalDef::Struct(_) => {
+                write!(f, "struct(..)")
+            }
+            NominalDef::Enum(_) => {
+                write!(f, "enum(..)")
+            }
+        }
```

Comment:
So, im guessing later these will have impls for fields of these types

Reply:
Could make a `verbose` mode or something of that nature

Reply:
Yeah, I'll add a @@Future comment for it.